### ISSUE TITLE
HashSet<enum>.Contains could not be translated

### ISSUE THREAD

**Issue by IgorMenshikov** [dotnet/efcore#17877]
I have a quite simple table named Operation and class for it. It has a property named Type mapped to an enum:

``` C#
[Column("OperationTypeID")]
public OperationType.Code Type { get; set; }
```

I try to request all operations with a filter by type and other fields:
``` C#
var list = (from i in context.Operation
                        where typeList.Contains(i.Type) &&
                            i.DeleteDate == null &&
                            i.PeriodID == parameters.PeriodID
                        orderby i.PlanDate descending
                        select new L.Item()
                        {
                            Operation = i
                        }).ToList();
```

List of necessary types is stored at the HashSet<OperationType.Code> typeList variable.

On 2.2. this query works OK. On 3.0.0-rc1 it fails with an error:

```
System.InvalidOperationException: 'The LINQ expression 'Where<Operation>(
    source: DbSet<Operation>, 
    predicate: (o) => HashSet<Code> { Submission, Reset, Calculate, ESIPaymentReport, Retraction, RetractionUndo, MoveRecord, MoveRecordUndo, CVSCAG, BIUpdate, }.Contains(o.Type) && o.DeleteDate == null && o.PeriodID == (Unhandled parameter: __parameters_PeriodID_0))' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync().
```

If I change HashSet to List, the query works.

Can HashSet be translated as List for queries?

### Further technical details

EF Core version: 3.0.0-rc1.19456.14
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.3 Preview 4

**Comment by smitpatel** [comment]
Possibly same root cause as #17342

**Comment by MurariAlex** [comment]
`HashSet<T>.Contains` implements `ICollection<T>.Contains` so it really is the same root cause as #17342.

This is fixed on 3.1.0 by #17599.

**Comment by IgorMenshikov** [comment]
thanks for the information. will wait till 3.1.0 to try again.

**Comment by Eirenarch** [comment]
@roji you've you've closed two issues as duplicates of each other, where do we track the state of the issue?

**Comment by roji** [comment]
The close event 30 minutes ago was just some automatic issue cleanup - the issue was already closed. I can indeed see that the two issues are flagged as duplicates of one another (not good), but there's an indication above that the issue was already fixed a long time ago - for 3.1.0. Are you saying you know there's still a problem here?

**Comment by Eirenarch** [comment]
Nope, I have not indication that there is an issue, I will have to try my particular problem. I just received a bunch of notifications from another issue and followed the chain until I got stuck in a loop :) I'll have to check my original issue but it might have been fixed at some point. Sorry to bother you
